src/plugins/explore/public/application/legacy/discover/application/components/visualizations/utils/use_debounced_value.ts

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { useState, useEffect, useRef } from 'react';
+
+/**
+ * Custom hook for debouncing values
+ * @param value The value to debounce
+ * @param onChange The callback to call with the debounced value
+ * @param delay The debounce delay in milliseconds
+ * @returns [localValue, setLocalValue] - The local value and setter for immediate UI updates
+ */
+export function useDebouncedValue<T>(
+  value: T,
+  onChange: (value: T) => void,
+  delay: number = 500
+): [T, (value: T) => void] {
+  const [localValue, setLocalValue] = useState<T>(value);
+  const timeoutRef = useRef<NodeJS.Timeout | null>(null);
```

Comment:
I think it should be `number` not `NodeJS.Timeout` in browser environment:
```suggestion
  const timeoutRef = useRef<number | null>(null);
```

Reply:
in package node_modules/@types/node/timers.d.ts, setTimeout type is NodeJS.Timeout. Changed it to number would give typescript error

```
function setTimeout<TArgs extends any[]>(
            callback: (...args: TArgs) => void,
            ms?: number,
            ...args: TArgs
        ): NodeJS.Timeout;

Reply:
Actually we can do `window.setTimeout()` which is should return a `number`